lib/mbedtls_errors/error.fmt

```diff
@@ -0,0 +1,165 @@
+/*
+ *  Error message information
+ *
+ *  Copyright (C) 2006-2015, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_ERROR_C) || defined(MBEDTLS_ERROR_STRERROR_DUMMY)
+#include "mbedtls/error.h"
+#include <string.h>
+#endif
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#define mbedtls_snprintf snprintf
+#define mbedtls_time_t   time_t
+#endif
+
+#if defined(MBEDTLS_ERROR_C)
+
+#include <stdio.h>
+
+HEADER_INCLUDED
+
+// Error code table type
+struct ssl_errs {
+    int16_t errnum;
+    const char *errstr;
+};
+
+// Table of high level error codes
+static const struct ssl_errs mbedtls_high_level_error_tab[] = {
+// BEGIN generated code
+HIGH_LEVEL_CODE_CHECKS
+// END generated code
+};
+
+static const struct ssl_errs mbedtls_low_level_error_tab[] = {
+// Low level error codes
+//
+// BEGIN generated code
+LOW_LEVEL_CODE_CHECKS
+// END generated code
+};
+
+static const char *mbedtls_err_prefix = "MBEDTLS_ERR_";
+#define MBEDTLS_ERR_PREFIX_LEN ( sizeof("MBEDTLS_ERR_")-1 )
```

Comment:
Is it worth having this prefix at all?  IMO the error messages are quite clear without it.  Not having it would save a small bit of flash in the routine below, and require less RAM to store the generated error message.  In fact, without the prefix would it be possible to just return a const pointer to the error string (like done in axtls)?

Reply:
I kept the prefix for two reasons. The first is that I did a couple of searches of error names with and without prefix, e.g.: https://www.google.com/search?q=MBEDTLS_ERR_CIPHER_AUTH_FAILED vs https://www.google.com/search?q=CIPHER_AUTH_FAILED and the results with the prefix seem noticeably better and lead straight to Mbedtls stuff while without the prefix you have to look down the page to find mbedtls.
The second reason is that because of the "upper" and "lower" errors that can get combined it's not always possible to return a const pointer into the table. I didn't want to dive in and make a special case for when that's possible (seems like the majority of cases, but I haven't investigated).
I thought that given the huge savings from the original mbedtls code it's OK to keep the prefix...

Reply:
Ah, I see, I didn't appreciate that there could be a combined error message from the upper and lower parts of the error code.  In that case it's not possible to make it return a const str.  So I guess the overhead of `"MBEDTLS_ERR_"` is ok.